### ISSUE TITLE
[rush] Improve command-line for "rush add"

### ISSUE THREAD

**Issue by qm3ster** [microsoft/rushstack#1160]
Not sure why `rush add` takes packages through the `-p` flag and not just naked parameters like all the package managers do, but it should still be able to install/update version of multiple package identifiers at once, for example with multiple flags: `-p PACKAGE -p PACKAGE@VERSION`.

The current behavior when you supply `rush add -p foo -p bar` is particularly surprising - instead of erroring out since this isn't supported, it just ignores the package names before the last one supplied. You end up only adding `bar`. That's probably a Bad Thing™

**Comment by octogonz** [comment]
It's a limitation of the **ts-command-line** library.  https://github.com/Microsoft/web-build-tools/issues/790 is a similar issue.  For a long time we've been meaning to replace the underlying parser that **ts-command-line** uses, but it hasn't been prioritized because we actually don't receive a lot of complaints about it compared to other things. :-)

**Comment by domarmstrong** [comment]
Can I add another complaint them :) this has been a pain for myself also. The -p flag seems pretty redundant. And having to rush add -p foo then wait, then add another, when you have 5 packages or so to add!

**Comment by octogonz** [comment]
> And having to rush add -p foo then wait, then add another, when you have 5 packages or so to add!

You can skip the waiting by using `--skip-update`, like this:

```bash
# Skip updates
$ rush add -p lib1 -s
$ rush add -p lib2 -s
$ rush add -p lib3 -s
$ rush add -p lib4 -s
# Perform one update for all 5 additions
$ rush add -p lib5
```

BTW without any changes to **ts-command-line**, we could convert `-p` into a string array, so you could do this:

```bash
$ rush add -p lib1 -p lib2 -p lib3 -p lib4 -p lib5
```

> The current behavior when you supply `rush add -p foo -p bar` is particularly surprising - instead of erroring out since this isn't supported, it just ignores the package names before the last one supplied. You end up only adding `bar`. That's probably a Bad Thing™

Damn... You guys are really making me want to fix the **ts-command-line** backend parser, though!  Today it's based on [argparse](https://www.npmjs.com/package/argparse) which hasn't been updated in over a year...

**Comment by slavafomin** [comment]
Well, this is extremely annoying now to add multiple packages through the `rush add` command. And we mostly add multiple packages, one for the library itself and another one for it's typings.

It looks like this:

```
rush add --caret --dev -p some-package --no-update
rush add --caret --dev -p @types/some-package
```

which is very cumbersome.

It would be nice to do just:

`rush add --caret --dev some-package @types/some-package`

Also, why `--dev` flag doesn't have a short version `-D` like in `npm`?

I believe it would be very beneficial to have a CLI syntax compatible to `npm`, for once it would help with migration and adaptation.

**Comment by Js-Brecht** [comment]
@octogonz What are your thoughts on using `defineCommandLineRemainder()` (#1869) to resolve this issue?  Do you think it could work?

**Comment by shedali** [comment]
running into this multiple times a day, it can be a bit jarring if having to switch between different package managers. Using bash function as workaround for now

```
rushadd(){
  for var arg in "$@";
  do rush add -p "$arg" --make-consistent;
  done;
}
```

**Comment by octogonz** [comment]
> What are your thoughts on using `defineCommandLineRemainder()` (#1869) to resolve this issue? Do you think it could work?

I'm not opposed to `$ rush add lib1 lib2 lib3 lib4 lib5`, but it's a new syntax that may require some design discussion.  (Also, as I recall after we implemented `defineCommandLineRemainder()` it turned out to have a parsing ambiguity, that made me want to rethink its design.  I don't remember the details offhand, though.)

If we want to get something implemented quickly, it would be simpler to extend the current design like this:

```
$ rush add -p lib1 -p lib2 -p lib3 -p lib4 -p lib5
```

That would just involve changing this:
https://github.com/microsoft/rushstack/blob/8c492e9c79c8d31acb9931694f6f7317eff187bf/apps/rush-lib/src/cli/actions/AddAction.ts#L48-L58
...to be a `CommandLineStringListParameter` instead of `CommandLineStringParameter`, and then add a for-loop somewhere to process each of the names.

Should be a pretty easy PR, and without any open questions for the design spec as far as I can tell. This would get something implemented quickly, and then we could come back and improve the CLI in a separate PR.
